great_expectations/types/base.py

```diff
@@ -27,138 +27,83 @@ def __dir__(self):
     def __deepcopy__(self, memo):
         return DotDict([(copy.deepcopy(k, memo), copy.deepcopy(v, memo)) for k, v in self.items()])
 
+    # The following are required to support yaml serialization, since we do not raise
+    # AttributeError from __getattr__ in DotDict. We *do* raise that AttributeError when it is possible to know
+    # a given attribute is not allowed (because it's not in _allowed_keys)
+    _yaml_merge = []
 
-# Inspiration : https://codereview.stackexchange.com/questions/81794/dictionary-with-restricted-keys
-class LooselyTypedDotDict(DotDict):
-    """dot.notation access to dictionary attributes, with limited keys
-
-
-    Note: this class is pretty useless on its own.
-    You need to subclass it like so:
+    @classmethod
+    def yaml_anchor(cls):
+        # This is required since our dotdict allows *any* access via dotNotation, blocking the normal
+        # behavior of raising an AttributeError when trying to access a nonexistent function
+        return None
 
-    class MyLooselyTypedDotDict(LooselyTypedDotDict):
-        _allowed_keys = set([
-            "x", "y", "z"
-        ])
+    @classmethod
+    def to_yaml(cls, representer, node):
+        """Use dict representation for DotDict (and subtypes by default)"""
+        return representer.represent_dict(node)
 
-    """
 
-    _allowed_keys = set()
+@yaml_object(yaml)
+class RequiredKeysDotDict(DotDict):
```

Comment:
I'm a little unclear on how `LooselyTypedDotDict` inherits from `RequiredKeysDotDict`, since in LTDD, `_allowed_keys` is required but `_required_keys` is optional. (Wow, that sentence is hard to scan.)

A docstring would help...

Reply:
Ha. Good point. RequiredKeysDotDict may just be a bad name...`_required_keys` isn't required, but it *can* be provided, but `_allowed_keys` cannot. (So you can optionally require certain keys, just like you can optionally specify types).

I'm updating docstrings to clarify